{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "sns.set_style('white')\n",
    "sns.set_style('ticks')\n",
    "import pandas as pd\n",
    "import zarr\n",
    "import allel; print('scikit-allel', allel.__version__)\n",
    "# check which version is installed\n",
    "\n",
    "#### Read VCF or zarr files\n",
    "chr = '2L'\n",
    "vcf_path = f'/scratch/daron_anopheles/anopheles-rose.{chr}.unifiedGenotyper.cov14x.vcf.gz'\n",
    "zarr_path = f'/scratch/daron_anopheles/anopheles-rose.{chr}.unifiedGenotyper.cov14x.zarr'\n",
    "if not os.path.isdir(zarr_path):\n",
    "    allel.vcf_to_zarr(vcf_path, zarr_path, fields='*', overwrite=True)\n",
    "callset = zarr.open_group(zarr_path, mode='r')\n",
    "\n",
    "vcf_path = f'/scratch/daron_anopheles/GAgam/ag1000g.phase2.ar1.pass.biallelic.{chr}.GAgam.vcf.gz'\n",
    "zarr_path = f'/scratch/daron_anopheles/GAgam/ag1000g.phase2.ar1.pass.biallelic.{chr}.GAgam.zarr'\n",
    "if not os.path.isdir(zarr_path):\n",
    "    allel.vcf_to_zarr(vcf_path, zarr_path, fields='*', overwrite=True)\n",
    "callset_GAgam = zarr.open_group(zarr_path, mode='r')\n",
    "\n",
    "vcf_path = f'/scratch/daron_anopheles/accessibility/accessibility.{chr}.vcf.gz'\n",
    "zarr_path = f'/scratch/daron_anopheles/accessibility/accessibility.{chr}.zarr'\n",
    "if not os.path.isdir(zarr_path):\n",
    "    allel.vcf_to_zarr(vcf_path, zarr_path, fields='*', overwrite=True)\n",
    "callset_access = zarr.open_group(zarr_path, mode='r')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Visualize variant attributes\n",
    "# https://gatk.broadinstitute.org/hc/en-us/articles/360035890471-Hard-filtering-germline-short-variants\n",
    "\n",
    "fig, ((ax0, ax1), (ax2, ax3), (ax4, ax5)) = plt.subplots(nrows=3, ncols=2, figsize=(15,8))\n",
    "#sns.despine(ax=ax, offset=10)\n",
    "\n",
    "# DP\n",
    "multi_x = (callset[\"variants/DP\"], callset_GAgam[\"variants/DP\"])\n",
    "# ax0.hist(multi_x, bins=range(0, 4000 + 50, 50), color=['C0', 'C2'])\n",
    "ax0.hist(multi_x, bins=100, color=['C1', '#38A055'], label=[\"Raw SNPs set\", \"GAgam\"])\n",
    "ax0.set_xlabel(\"DP\")\n",
    "ax0.set_ylabel('No. variants')\n",
    "ax0.set_title('Variant DP distribution')\n",
    "ax0.legend(loc='upper right')\n",
    "\n",
    "# FS fisher strand\n",
    "multi_x = (callset[\"variants/FS\"], callset_GAgam[\"variants/FS\"])\n",
    "ax1.hist(multi_x, bins=range(0, 100 + 5, 5), color=['C1', '#38A055'])\n",
    "ax1.set_xlabel(\"FS\")\n",
    "ax1.set_ylabel('No. variants')\n",
    "ax1.set_title('Variant FS distribution')\n",
    "ax1.axvline(x=60, c='black')\n",
    "\n",
    "# QD qualbydepth\n",
    "multi_x = (callset[\"variants/QD\"], callset_GAgam[\"variants/QD\"])\n",
    "ax2.hist(multi_x, bins=40, color=['C1', '#38A055'])\n",
    "ax2.set_xlabel(\"QD\")\n",
    "ax2.set_ylabel('No. variants')\n",
    "ax2.set_title('Variant QD distribution')\n",
    "ax2.axvline(x=2, c='black')\n",
    "ax2.axvline(x=5, c='black')\n",
    "\n",
    "# ReadPosRankSum\n",
    "x = np.array(callset[\"variants/ReadPosRankSum\"])\n",
    "x_no_nan = x[~np.isnan(x)]\n",
    "x = np.array(callset_GAgam[\"variants/ReadPosRankSum\"])\n",
    "x_GAgam_no_nan = x[~np.isnan(x)]\n",
    "multi_x = (x_no_nan, x_GAgam_no_nan)\n",
    "ax3.hist(multi_x, bins=100, color=['C1', '#38A055'])\n",
    "ax3.set_xlabel(\"ReadPosRankSum\")\n",
    "ax3.set_ylabel('No. variants')\n",
    "ax3.set_title('Variant ReadPosRankSum distribution')\n",
    "ax3.axvline(x=-8, c='black')\n",
    "\n",
    "# Number of alternative allele\n",
    "multi_x = (callset[\"variants/numalt\"], callset_GAgam[\"variants/numalt\"])\n",
    "ax4.hist(multi_x, bins=np.arange(0.5, 5.5, 1), color=['C1', '#38A055'])\n",
    "ax4.set_xlabel(\"numalt\")\n",
    "ax4.set_ylabel('No. variants')\n",
    "ax4.set_title('Variant numalt distribution')\n",
    "\n",
    "# Distribution missing data\n",
    "miss = np.count_nonzero(callset[\"calldata/GT\"][:,:,0] == -1, axis=1)\n",
    "miss_pct = (miss/callset[\"calldata/GT\"].shape[1])*100\n",
    "miss_hist, bin_edges = np.histogram(miss_pct, bins=range(0, 50 + 2, 2))\n",
    "miss_hist = (miss_hist/miss_hist.sum())*100\n",
    "\n",
    "miss_pass = np.count_nonzero(callset_GAgam[\"calldata/GT\"][:,:,0] == -1, axis=1)\n",
    "miss_pass_pct = (miss_pass/callset_GAgam[\"calldata/GT\"].shape[1])*100\n",
    "miss_pass_hist, bin_edges = np.histogram(miss_pass_pct, bins=range(0, 50 + 2, 2))\n",
    "miss_pass_hist = (miss_pass_hist/miss_pass_hist.sum())*100\n",
    "\n",
    "bar_width = 1\n",
    "ax5.bar(bin_edges[range(1,len(miss_hist)+1)] - (bar_width/2), miss_hist, width=bar_width, color='C1')\n",
    "ax5.bar(bin_edges[range(1,len(miss_hist)+1)] + (bar_width/2), miss_pass_hist, width=bar_width, color='#38A055')\n",
    "ax5.set_xlabel(\"Missing data in proportion at each site\")\n",
    "ax5.set_ylabel('No. variants in %')\n",
    "ax5.set_title('Missingness per variants')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Filtration of variant\n",
    "fig, ((ax0, ax1)) = plt.subplots(nrows=1, ncols=2, figsize=(15,5))\n",
    "\n",
    "#### Accessibility filtering\n",
    "\n",
    "access = np.array(callset_access['variants/POS'])[np.array(callset_access['variants/Accessible'])]\n",
    "intersection = np.intersect1d(access, np.array(callset['variants/POS']))\n",
    "access_pct = 100-(len(intersection)/len(np.array(callset['variants/POS']))*100)\n",
    "pass_snp_pos = intersection\n",
    "\n",
    "#### hard filtering\n",
    "## FS\n",
    "array = np.array(callset['variants/FS'])\n",
    "FS_pct = (len(array[array>60])/len(array))*100\n",
    "pos_pass = np.array(callset['variants/POS'])[array<=60]\n",
    "pass_snp_pos = np.intersect1d(pos_pass, pass_snp_pos)\n",
    "\n",
    "## ReadPosRankSum\n",
    "array = np.array(callset['variants/ReadPosRankSum'])\n",
    "array_no_nan = array[~np.isnan(array)]\n",
    "ReadPosRankSum_pct = (len(array_no_nan[array_no_nan<-8.000])/len(array_no_nan))*100\n",
    "pos_pass = np.array(callset['variants/POS'])[array>=-8.000]\n",
    "pass_snp_pos = np.intersect1d(pos_pass, pass_snp_pos)\n",
    "\n",
    "## QD\n",
    "array = np.array(callset['variants/QD'])\n",
    "QD_pct_5 = (len(array[array<5])/len(array))*100\n",
    "QD_pct_2 = (len(array[array<2])/len(array))*100\n",
    "\n",
    "pos_pass = np.array(callset['variants/POS'])[array>=5]\n",
    "pass_snp_pos_QD_5 = np.intersect1d(pos_pass, pass_snp_pos)\n",
    "\n",
    "pos_pass = np.array(callset['variants/POS'])[array>=2]\n",
    "pass_snp_pos_QD_2 = np.intersect1d(pos_pass, pass_snp_pos)\n",
    "\n",
    "print(\"Number of SNP passing QD>2:\", len(pass_snp_pos_QD_2))\n",
    "print(\"Number of SNP passing QD>5:\", len(pass_snp_pos_QD_5))\n",
    "\n",
    "#### Plot \n",
    "bar_width = 0.5\n",
    "y = [len(callset_GAgam['variants/POS']), len(callset['variants/POS']),len(pass_snp_pos_QD_2), len(pass_snp_pos_QD_5)]\n",
    "x = ['GAgam SNPs', 'Raw SNPs', 'Filered SNPs\\n(QD<2)', 'Filered SNPs\\n(QD<5)']\n",
    "ax0.bar(np.arange(len(x)), y, bar_width, color=['#38A055', 'C1', 'C0', 'C3'])\n",
    "ax0.set_xticks(np.arange(len(x)))\n",
    "ax0.set_xticklabels(x)\n",
    "ax0.set_xlabel('')\n",
    "ax0.set_ylabel('# SNPs')\n",
    "ax0.set_title('Total number of SNP')\n",
    "\n",
    "x = ['Accessibility', 'QD_2', 'QD_5', 'FS', 'ReadPosRankSum']\n",
    "y = [access_pct, QD_pct_2, QD_pct_5, FS_pct, ReadPosRankSum_pct]\n",
    "ax1.bar(np.arange(len(x)), y, bar_width, color=['c', 'b', 'b', 'y', 'r'])\n",
    "ax1.set_xticks(np.arange(len(x)))\n",
    "ax1.set_xticklabels(x)\n",
    "ax1.set_xlabel('')\n",
    "ax1.set_ylabel('% of SNPs')\n",
    "ax1.set_title('Proportion of SNPs filtered per category')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Visualize variant density\n",
    "pos = callset['variants/POS']\n",
    "pos_pass = pass_snp_pos_QD_5\n",
    "pos_GAgam = callset_GAgam['variants/POS']\n",
    "\n",
    "pos_max = 0\n",
    "if max(pos) < max(pos_pass):\n",
    "    pos_max = max(pos_pass)\n",
    "else:\n",
    "    pos_max = max(pos)\n",
    "\n",
    "window_size = 50000\n",
    "bins = np.arange(0, pos_max, window_size)\n",
    "\n",
    "# use window midpoints as x coordinate\n",
    "x = (bins[1:] + bins[:-1])/2\n",
    "\n",
    "# compute variant density in each window\n",
    "h, _ = np.histogram(pos, bins=bins)\n",
    "y = h / window_size\n",
    "h, _ = np.histogram(pos_pass, bins=bins)\n",
    "y_pass = h / window_size\n",
    "h, _ = np.histogram(pos_GAgam, bins=bins)\n",
    "y_GAgam = h / window_size\n",
    "\n",
    "# plot\n",
    "fig, ax = plt.subplots(figsize=(15, 3))\n",
    "# sns.despine(ax=ax, offset=10)\n",
    "ax.plot(x, y, linestyle=\":\", label=\"Raw SNPs\", color=\"C1\")\n",
    "ax.plot(x, y_pass, label=\"Filtered SNPs (QD<5)\", color=\"C3\")\n",
    "ax.plot(x, y_GAgam, label=\"GAgam SNPs\", color=\"#38A055\")\n",
    "ax.legend()\n",
    "ax.set_xlabel('Chromosome position (bp)')\n",
    "ax.set_ylabel('Variant density (bp$^{-1}$)')\n",
    "ax.set_title('variant density')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Visualize calldata attributes GQ and DP\n",
    "pos = allel.SortedIndex(callset['variants/POS'])\n",
    "index = np.in1d(pos, pass_snp_pos_QD_5).nonzero()[0]\n",
    "\n",
    "loc_variant_selection, _ = pos.locate_intersection(pass_snp_pos_QD_5)\n",
    "gt = allel.GenotypeArray(callset['calldata/GT'])\n",
    "gt_variant_selection = gt.compress(loc_variant_selection, axis=0)\n",
    "\n",
    "gq = np.array(callset[\"calldata/GQ\"])\n",
    "gq_variant_selection = gq[index]\n",
    "\n",
    "dp = np.array(callset[\"calldata/DP\"])\n",
    "dp_variant_selection = dp[index]\n",
    "\n",
    "# raise SystemExit(\"Stop right there!\")\n",
    "\n",
    "gt_GAgam_zarr = allel.GenotypeArray(callset_GAgam[\"calldata/GT\"])\n",
    "gq_GAgam = np.array(callset_GAgam[\"calldata/GQ\"])\n",
    "dp_GAgam = np.array(callset_GAgam[\"calldata/DP\"])\n",
    "\n",
    "####\n",
    "fig, ((ax0, ax1)) = plt.subplots(nrows=1, ncols=2, figsize=(15,5))\n",
    "\n",
    "# GQ\n",
    "# multi_x = (callset[\"calldata/GQ\"][~np.isin(callset[\"calldata/GT\"][:,:,0], -1)], callset_pass[\"calldata/GQ\"][~np.isin(callset_pass[\"calldata/GT\"][:,:,0], -1)])\n",
    "multi_x = (gq_variant_selection[~gt_variant_selection.is_missing()], gq_GAgam[~gt_GAgam_zarr.is_missing()])\n",
    "ax0.hist(multi_x, bins=30, color=[\"C3\", \"#38A055\"], label=[\"Filtered SNPs (QD<5)\", \"GAgam\"])\n",
    "ax0.set_xlabel(\"GQ\")\n",
    "ax0.set_ylabel('No. Genotypes')\n",
    "ax0.set_title('Genotype GQ distribution')\n",
    "ax0.legend()\n",
    "\n",
    "# DP\n",
    "#multi_x = (callset[\"calldata/DP\"][~np.isin(callset[\"calldata/GT\"][:,:,0], -1)], callset_pass[\"calldata/DP\"][~np.isin(callset_pass[\"calldata/GT\"][:,:,0], -1)])\n",
    "multi_x = (dp_variant_selection[~gt_variant_selection.is_missing()], dp_GAgam[~gt_GAgam_zarr.is_missing()])\n",
    "\n",
    "# ax1.hist(multi_x, bins=50)\n",
    "ax1.hist(multi_x, bins=range(0, 100 + 5, 5), color=[\"C3\", \"#38A055\"], label=[\"Filtered SNPs (QD<5)\", \"GAgam\"])\n",
    "ax1.set_xlabel(\"DP\")\n",
    "ax1.set_ylabel('No. Genotypes')\n",
    "ax1.set_title('Genotype DP distribution')\n",
    "ax1.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#### Visualize distribution of missing data per variants \n",
    "fig, ((ax0, ax1, ax2)) = plt.subplots(nrows=1, ncols=3, figsize=(15,5))\n",
    "\n",
    "# GQ < 15\n",
    "miss = np.count_nonzero(gq_variant_selection < 15, axis=1)\n",
    "miss_pct = (miss/gq_variant_selection.shape[1])*100\n",
    "lmiss_gq15_5 = (len(miss_pct[miss_pct<5]))\n",
    "lmiss_gq15_10 = (len(miss_pct[miss_pct<10]))\n",
    "\n",
    "ax0.hist(miss_pct, bins=range(0, 50 + 1, 1), color=\"C0\")\n",
    "ax0.set_xlabel(\"Percentage of missing data per variant\")\n",
    "ax0.set_ylabel('No. Variants')\n",
    "ax0.set_title('GQ<15')\n",
    "ax0.axvline(x=5, c='black')\n",
    "ax0.axvline(x=10, c='black')\n",
    "\n",
    "# GQ < 20\n",
    "miss = np.count_nonzero(gq_variant_selection < 20, axis=1)\n",
    "miss_pct = (miss/gq_variant_selection.shape[1])*100\n",
    "lmiss_gq20_5 = (len(miss_pct[miss_pct<5]))\n",
    "lmiss_gq20_10 = (len(miss_pct[miss_pct<10]))\n",
    "\n",
    "ax1.hist(miss_pct, bins=range(0, 50 + 2, 1), color=\"C2\")\n",
    "ax1.set_xlabel(\"Percentage of missing data per variant\")\n",
    "ax1.set_ylabel('No. Variants')\n",
    "ax1.set_title('GQ<20')\n",
    "ax1.axvline(x=5, c='black')\n",
    "ax1.axvline(x=10, c='black')\n",
    "\n",
    "# print(len(miss_pct))\n",
    "## Proportion of SNP removed \n",
    "\n",
    "y = [len(callset_GAgam['variants/POS']), len(callset['variants/POS']), gq_variant_selection.shape[0], lmiss_gq15_5, lmiss_gq15_10, lmiss_gq20_5, lmiss_gq20_10]\n",
    "x = ['GAgam SNPs', 'raw SNPs', 'QD<5 SNPs', 'GQ<15\\nlmiss<5%', 'GQ<15\\nlmiss<10%', 'GQ<20\\nlmiss<5%', 'GQ<20\\nlmiss<10%']\n",
    "ax2.bar(np.arange(len(x)), y, bar_width, color=['#38A055', 'C1', 'C0', 'C8', 'C6', 'C3', 'C9'])\n",
    "ax2.set_xticks(np.arange(len(x)))\n",
    "ax2.set_xticklabels(x, rotation=45)\n",
    "ax2.set_xlabel('')\n",
    "ax2.set_ylabel('No. of SNPs')\n",
    "ax2.set_title('Overall No. of SNPs')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "miss = np.count_nonzero(gq_variant_selection < 20, axis=1)\n",
    "miss_pct = (miss/gq_variant_selection.shape[1])*100\n",
    "gt_variant_selection_lmiss = gt_variant_selection[(miss_pct<5)]\n",
    "gq_variant_selection_lmiss = gq_variant_selection[(miss_pct<5)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Filtering Individual based on proportion of missing data\n",
    "\n",
    "fig, ((ax0, ax1, ax2), (ax3, ax4, ax5)) = plt.subplots(nrows=2, ncols=3, figsize=(15,5))\n",
    "\n",
    "count_imiss_rawSNPs = np.count_nonzero(gt.is_missing() == True, axis=0)\n",
    "pct_imiss_rawSNPs = (count_imiss_rawSNPs/gt.shape[0])*100\n",
    "ax0.bar(range(0, len(pct_imiss_rawSNPs)), pct_imiss_rawSNPs, color=\"C1\")\n",
    "ax0.set_ylabel(\"Percentage of missing data\")\n",
    "ax0.set_xlabel('Individual index')\n",
    "ax0.set_title('Individual missingness raw SNPs set')\n",
    "ax0.set_ylim([0,20])\n",
    "\n",
    "count_imiss_variant_selection = np.count_nonzero((gq_variant_selection < 20), axis=0) # count GQ bellow threshold\n",
    "pct_imiss_variant_selection = (count_imiss_variant_selection/gt_variant_selection.shape[0])*100\n",
    "ax1.bar(range(0, len(pct_imiss_variant_selection)), pct_imiss_variant_selection, color=\"C0\")\n",
    "ax1.set_ylabel(\"Percentage of missing data\")\n",
    "ax1.set_xlabel('Individual index')\n",
    "ax1.set_title('Individual missingness QD<5 GQ<20')\n",
    "ax1.set_ylim([0,20])\n",
    "\n",
    "count_imiss_variant_selection_lmiss = np.count_nonzero((gq_variant_selection_lmiss < 20), axis=0) # count GQ bellow threshold\n",
    "pct_imiss_variant_selection_lmiss = (count_imiss_variant_selection_lmiss/gq_variant_selection_lmiss.shape[0])*100\n",
    "ax2.bar(range(0, len(pct_imiss_variant_selection_lmiss)), pct_imiss_variant_selection_lmiss, color=\"C3\")\n",
    "ax2.set_ylabel(\"Percentage of missing data\")\n",
    "ax2.set_xlabel('Individual index')\n",
    "ax2.set_title('Individual missingness QD<5 GQ<20 lmiss<5%')\n",
    "ax2.set_ylim([0,20])\n",
    "ax2.axhline(y=5, c='black')\n",
    "ax2.axhline(y=10, c='black')\n",
    "\n",
    "#### distribution missigness individual\n",
    "multi_x = (pct_imiss_rawSNPs, pct_imiss_variant_selection, pct_imiss_variant_selection_lmiss)\n",
    "ax3.hist(multi_x, bins=range(0, 100 + 5, 5), color=[\"C1\", \"C0\", \"C3\"], label=[\"Raw SNPs\", \"QD<5\", \"QD<5 GQ<20 lmiss<5%\"])\n",
    "ax3.set_xlabel(\"% of missing data\")\n",
    "ax3.set_ylabel('No. of individual')\n",
    "ax3.legend()\n",
    "ax3.set_title('Distribution of individual missingness')\n",
    "\n",
    "## nb individual to keep \n",
    "x = ['QD<5\\nGQ<20\\nlmiss<5%\\nimiss5%', 'QD<5\\nGQ<20\\nlmiss<5%\\nimiss10%']\n",
    "y = [gq_variant_selection_lmiss.shape[1]-len(pct_imiss_variant_selection_lmiss[pct_imiss_variant_selection_lmiss<5]),\n",
    "     gq_variant_selection_lmiss.shape[1]-len(pct_imiss_variant_selection_lmiss[pct_imiss_variant_selection_lmiss<10])]\n",
    "print(y)\n",
    "ax4.bar(np.arange(len(x)), y, bar_width, color=['C6', 'C9'])\n",
    "ax4.set_xticks(np.arange(len(x)))\n",
    "ax4.set_xticklabels(x, rotation=0)\n",
    "ax4.set_xlabel('')\n",
    "ax4.set_ylabel('No. of individual')\n",
    "ax4.set_title('No. of individual to discard')\n",
    "\n",
    "## Overall proportion of missing data\n",
    "lmiss_thresold = 5\n",
    "gq_variant_selection_lmiss_imiss = gq_variant_selection_lmiss[:,pct_imiss_variant_selection_lmiss < lmiss_thresold] \n",
    "# gt_variant_selection_lmiss_imiss = gt_variant_selection_lmiss[:,pct_imiss_variant_selection_lmiss < lmiss_thresold]\n",
    "\n",
    "count_imiss_variant_selection_lmiss_imiss = np.count_nonzero(gq_variant_selection_lmiss_imiss < 20, axis=0)\n",
    "\n",
    "y = [(count_imiss_rawSNPs.sum()/gt.size)*100,\n",
    "     (count_imiss_variant_selection.sum()/gt_variant_selection.size)*100,\n",
    "     (count_imiss_variant_selection_lmiss.sum()/gt_variant_selection_lmiss.size)*100,\n",
    "     (count_imiss_variant_selection_lmiss_imiss.sum()/gq_variant_selection_lmiss_imiss.size)*100]\n",
    "\n",
    "x = ['Raw SNPs', 'QD<5\\nGQ<20', 'QD<5\\nGQ<20\\nlmiss<5%', 'QD<5\\nGQ<20\\nlmiss<5%\\nimiss<5%']\n",
    "bar_width = 0.5\n",
    "ax5.bar(np.arange(len(x)), y, bar_width, color=['C1', 'C0', 'C3'])\n",
    "ax5.set_xticks(np.arange(len(x)))\n",
    "ax5.set_xticklabels(x)\n",
    "ax5.set_xlabel('')\n",
    "ax5.set_ylabel('% of missing data')\n",
    "ax5.set_title('Overall % of missing data')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_fn = '/data3/projects/plasmodium/anopheles/ressources/wilding_samples_info.txt'\n",
    "samplesInfo = pd.read_csv(samples_fn, encoding='utf-8-sig', sep='\\t')\n",
    "#samplesInfo.groupby([\"location\", \"site\"]).count()\n",
    "\n",
    "keepSamples = callset[\"/samples\"][:][pct_imiss_variant_selection_lmiss < 10]\n",
    "index = np.in1d(np.array(samplesInfo.id), keepSamples).nonzero()[0]\n",
    "samplesInfoSub = samplesInfo.iloc[index,:]\n",
    "\n",
    "fig, (ax0, ax1, ax2) = plt.subplots(nrows=1, ncols=3, figsize=(15,5))\n",
    "samplesInfoSub.groupby([\"population\"]).count()['id'].plot.bar(ax=ax0, rot=45)\n",
    "ax0.set_ylabel(\"No. Individual\")\n",
    "samplesInfoSub.groupby([\"collection_method\"]).count()['id'].plot.bar(ax=ax1, rot=45)\n",
    "samplesInfoSub.groupby([\"animal_sex\"]).count()['id'].plot.bar(ax=ax2, rot=45)\n",
    "\n",
    "print(samplesInfoSub.groupby([\"population\", \"animal_sex\"]).count()['id'])\n",
    "print(\"\\n\")\n",
    "nb = len(callset[\"/samples\"][:][pct_imiss_variant_selection_lmiss > 10])\n",
    "print(f\"{nb} samples will be removed because imiss > 10%:\")\n",
    "print(callset[\"/samples\"][:][pct_imiss_variant_selection_lmiss > 10])\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
